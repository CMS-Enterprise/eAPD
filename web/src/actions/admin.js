import { notify } from './notification';
import axios from '../util/api';

export const ADMIN_CREATE_USER_REQUEST = Symbol(
  'admin : create user : request'
);
export const ADMIN_CREATE_USER_SUCCESS = Symbol(
  'admin : create user : success'
);
export const ADMIN_CREATE_USER_ERROR = Symbol('admin : create user : error');

export const createUser = ({ email, name, password, state }) => dispatch => {
  dispatch({ type: ADMIN_CREATE_USER_REQUEST });

  return axios
    .post('/users', { email, name, password, state })
    .then(() => {
      dispatch({ type: ADMIN_CREATE_USER_SUCCESS });
      dispatch(notify('User created!'));
    })
    .catch(e => {
      dispatch({ type: ADMIN_CREATE_USER_ERROR });

      switch (e.response.data.error) {
        case 'add-user-invalid':
          dispatch(notify('Error: Email and password are required'));
          break;
        case 'add-user-email-exists':
          dispatch(notify('Error: A user with this email already exists'));
          break;
        case 'add-user-weak-password':
          dispatch(notify('Error: The provided password is too weak'));
          break;
        case 'add-user-invalid-phone':
          dispatch(notify('Error: The provided phone number is invalid'));
          break;
        default:
          dispatch(notify('Unknown error creating user'));
          break;
      }

      return Promise.reject();
    });
};
