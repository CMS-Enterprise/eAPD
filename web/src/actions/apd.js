import { push } from 'connected-react-router';

import { getIsAdmin } from '../reducers/user.selector';
import axios from '../util/api';
import { fromAPI, toAPI } from '../util/serialization/apd';

const LAST_APD_ID_STORAGE_KEY = 'last-apd-id';

export const ADD_APD_KEY_PERSON = 'ADD_APD_KEY_PERSON';
export const CREATE_APD = 'CREATE_APD';
export const CREATE_APD_REQUEST = 'CREATE_APD_REQUEST';
export const CREATE_APD_SUCCESS = 'CREATE_APD_SUCCESS';
export const CREATE_APD_FAILURE = 'CREATE_APD_FAILURE';
export const DELETE_APD_REQUEST = Symbol('delete apd : request');
export const DELETE_APD_SUCCESS = Symbol('delete apd : success');
export const DELETE_APD_FAILURE = Symbol('delete apd : failure');
export const GET_APD_REQUEST = 'GET_APD_REQUEST';
export const GET_APD_SUCCESS = 'GET_APD_SUCCESS';
export const GET_APD_FAILURE = 'GET_APD_FAILURE';
export const REMOVE_APD_KEY_PERSON = 'REMOVE_APD_KEY_PERSON';
export const SAVE_APD_REQUEST = 'SAVE_APD_REQUEST';
export const SAVE_APD_SUCCESS = 'SAVE_APD_SUCCESS';
export const SAVE_APD_FAILURE = 'SAVE_APD_FAILURE';
export const SELECT_APD = 'SELECT_APD';
export const SUBMIT_APD_REQUEST = 'SUBMIT_APD_REQUEST';
export const SUBMIT_APD_SUCCESS = 'SUBMIT_APD_SUCCESS';
export const SUBMIT_APD_FAILURE = 'SUBMIT_APD_FAILURE';
export const UPDATE_APD = 'UPDATE_APD';
export const UPDATE_BUDGET = 'UPDATE_BUDGET';
export const WITHDRAW_APD_REQUEST = Symbol('withdraw apd request');
export const WITHDRAW_APD_SUCCESS = Symbol('withdraw apd success');
export const WITHDRAW_APD_FAILURE = Symbol('withdraw apd failure');

export const SET_SELECT_APD_ON_LOAD = 'SET_SELECT_APD_ON_LOAD';
export const selectApdOnLoad = () => (dispatch, getState) => {
  const isAdmin = getIsAdmin(getState());
  if (!isAdmin) {
    dispatch({ type: SET_SELECT_APD_ON_LOAD });
  }
};

export const addKeyPerson = () => ({ type: ADD_APD_KEY_PERSON });
export const removeKeyPerson = (key, { global = window } = {}) => dispatch => {
  if (global.confirm('Do you really want to delete this key person?')) {
    dispatch({
      type: REMOVE_APD_KEY_PERSON,
      key
    });
  }
};

export const updateBudget = () => (dispatch, getState) =>
  dispatch({ type: UPDATE_BUDGET, state: getState() });

export const requestApd = () => ({ type: GET_APD_REQUEST });
export const receiveApd = data => ({ type: GET_APD_SUCCESS, data });
export const failApd = error => ({ type: GET_APD_FAILURE, error });

export const updateApd = updates => dispatch => {
  dispatch({ type: UPDATE_APD, updates });
  if (updates.years) {
    dispatch(updateBudget());
  }
};

export const selectApd = (
  id,
  { deserialize = fromAPI, global = window, pushRoute = push } = {}
) => dispatch =>
  axios.get(`/apds/${id}`).then(req => {
    dispatch({ type: SELECT_APD, apd: deserialize(req.data) });
    dispatch(updateBudget());
    dispatch(pushRoute('/apd'));

    if (global.localStorage) {
      global.localStorage.setItem(LAST_APD_ID_STORAGE_KEY, id);
    }
  });

export const createRequest = () => ({ type: CREATE_APD_REQUEST });
export const createSuccess = data => ({ type: CREATE_APD_SUCCESS, data });
export const createFailure = () => ({ type: CREATE_APD_FAILURE });
export const createApd = ({
  deserialize = fromAPI,
  pushRoute = push
} = {}) => dispatch => {
  dispatch(createRequest());
  return axios
    .post('/apds')
    .then(async req => {
      const apd = deserialize(req.data);
      dispatch(createSuccess(apd));
      await dispatch(selectApd(apd.id, { deserialize, pushRoute }));
    })
    .catch(error => {
      const reason = error.response ? error.response.data : 'N/A';
      dispatch(createFailure(reason));
    });
};

export const requestSave = () => ({ type: SAVE_APD_REQUEST });
export const saveSuccess = data => ({ type: SAVE_APD_SUCCESS, data });
export const saveFailure = data => ({ type: SAVE_APD_FAILURE, data });

export const fetchApd = ({
  global = window,
  pushRoute = push,
  select = selectApd
} = {}) => (dispatch, getState) => {
  dispatch(requestApd());

  const url = `/apds`;

  return axios
    .get(url)
    .then(async req => {
      const apd = Array.isArray(req.data) ? req.data : null;
      dispatch(receiveApd(apd.filter(({ status }) => status !== 'archived')));

      const {
        apd: { selectAPDOnLoad }
      } = getState();
      if (
        selectAPDOnLoad &&
        global.localStorage &&
        global.localStorage.getItem(LAST_APD_ID_STORAGE_KEY)
      ) {
        await dispatch(
          select(global.localStorage.getItem('last-apd-id'), {
            global,
            pushRoute
          })
        );
      }
    })
    .catch(error => {
      const reason = error.response ? error.response.data : 'N/A';
      dispatch(failApd(reason));
    });
};

const shouldFetchApd = state => !state.apd.loaded;

export const fetchApdDataIfNeeded = () => (dispatch, getState) => {
  if (shouldFetchApd(getState())) {
    return dispatch(fetchApd());
  }

  return null;
};

export const saveApd = ({ serialize = toAPI } = {}) => (dispatch, state) => {
  const {
    apd: { data: updatedApd },
    activities,
    dirty
  } = state();

  if (!dirty.dirty) {
    return Promise.resolve();
  }

  dispatch(requestSave());

  const apd = serialize(updatedApd, activities);

  // These are the fields we want to remove if they aren't dirty.
  const filterAPDFields = [
    'federalCitations',
    'incentivePayments',
    'narrativeHIE',
    'narrativeHIT',
    'narrativeMMIS',
    'programOverview',
    'previousActivityExpenses',
    'previousActivitySummary',
    'stateProfile'
  ];
  const filterActivityFields = [
    'contractorResources',
    'costAllocation',
    'costAllocationNarrative',
    'expenses',
    'goals',
    'schedule',
    'standardsAndConditions',
    'statePersonnel',
    'quarterlyFFP'
  ];

  filterAPDFields.forEach(field => {
    if (!dirty.data.apd[field]) {
      delete apd[field];
    }
  });

  for (let i = apd.activities.length - 1; i >= 0; i -= 1) {
    const activity = apd.activities[i];
    if (!dirty.data.activities.byKey[activity.key]) {
      // If this activity isn't dirty, strip it from the list to save,
      // but then add just its ID back - this prevents it from being
      // deleted by the API synchronization.
      apd.activities.splice(i, 1);
      apd.activities.push({ id: activity.id });
    } else {
      const dirtyActivity = dirty.data.activities.byKey[activity.key];
      // Get rid of any activity fields that aren't dirty.
      filterActivityFields.forEach(field => {
        if (!dirtyActivity[field]) {
          delete activity[field];
        }
      });
    }
  }

  return axios
    .put(`/apds/${updatedApd.id}`, apd)
    .then(res => {
      dispatch(saveSuccess(res.data));
      return res.data;
    })
    .catch(error => {
      if (error.response.status === 403) {
        dispatch(saveFailure('save-apd.not-logged-in'));
      } else {
        dispatch(saveFailure());
      }
      throw error;
    });
};

export const deleteApd = (id, { fetch = fetchApd } = {}) => dispatch => {
  dispatch({ type: DELETE_APD_REQUEST });

  return axios
    .delete(`/apds/${id}`)
    .then(() => {
      dispatch({ type: DELETE_APD_SUCCESS });
      dispatch(fetch());
    })
    .catch(() => {
      dispatch({ type: DELETE_APD_FAILURE });
    });
};
