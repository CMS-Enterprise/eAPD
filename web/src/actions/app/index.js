export * from './apd';
export * from './symbols';
