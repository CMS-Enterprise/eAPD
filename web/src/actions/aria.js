export const ARIA_ANNOUNCE_CHANGE = Symbol('aria change announcement');

export const ariaAnnounceFFPQuarterly = (aKey, year, q, name) => (
  dispatch,
  getState
) => {
  dispatch({
    type: ARIA_ANNOUNCE_CHANGE,
    message: getState().budget.activities[aKey].quarterlyFFP[year][q][name]
      .dollars
  });
};

export const ariaAnnounceApdLoading = () => ({
  type: ARIA_ANNOUNCE_CHANGE,
  message: 'Your APD is loading'
});

export const ariaAnnounceApdLoaded = () => ({
  type: ARIA_ANNOUNCE_CHANGE,
  message: 'Your APD is loaded and ready to edit'
});
