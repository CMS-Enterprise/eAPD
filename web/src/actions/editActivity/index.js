export * from './contractorResources';
export * from './costAllocate';
export * from './goals';
export * from './overview';
