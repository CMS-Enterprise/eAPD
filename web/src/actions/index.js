export const INCREMENT_COUNT = 'INCREMENT_COUNT';
export const DECREMENT_COUNT = 'DECREMENT_COUNT';

export const increment = () => ({ type: INCREMENT_COUNT });
export const decrement = () => ({ type: DECREMENT_COUNT });
