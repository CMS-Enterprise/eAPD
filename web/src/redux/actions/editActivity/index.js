export * from './activityList';
export * from './conditionsForEnhancedFunding';
export * from './alternativesAndRisks';
export * from './contractorResources';
export * from './costAllocate';
export * from './nonPersonnelCosts';
export * from './outcomes';
export * from './overview';
export * from './scheduleAndMilestones';
export * from './standardsAndConditions';
export * from './statePersonnel';
