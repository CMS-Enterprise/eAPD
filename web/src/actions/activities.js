import { saveApd } from './apd';
import { updateBudget } from './budget';

export const ADD_ACTIVITY = 'ADD_ACTIVITY';
export const ADD_ACTIVITY_CONTRACTOR = 'ADD_ACTIVITY_CONTRACTOR';
export const ADD_ACTIVITY_DIRTY = 'ADD_ACTIVITY_DIRTY';
export const ADD_ACTIVITY_GOAL = 'ADD_ACTIVITY_GOAL';
export const ADD_ACTIVITY_EXPENSE = 'ADD_ACTIVITY_EXPENSE';
export const ADD_ACTIVITY_MILESTONE = 'ADD_ACTIVITY_MILESTONE';
export const ADD_ACTIVITY_STATE_PERSON = 'ADD_ACTIVITY_STATE_PERSON';
export const EXPAND_ACTIVITY_SECTION = 'EXPAND_ACTIVITY_SECTION';
export const REMOVE_ACTIVITY = 'REMOVE_ACTIVITY';
export const REMOVE_ACTIVITY_CONTRACTOR = 'REMOVE_ACTIVITY_CONTRACTOR';
export const REMOVE_ACTIVITY_GOAL = 'REMOVE_ACTIVITY_GOAL';
export const REMOVE_ACTIVITY_EXPENSE = 'REMOVE_ACTIVITY_EXPENSE';
export const REMOVE_ACTIVITY_MILESTONE = 'REMOVE_ACTIVITY_MILESTONE';
export const REMOVE_ACTIVITY_STATE_PERSON = 'REMOVE_ACTIVITY_STATE_PERSON';
export const TOGGLE_ACTIVITY_CONTRACTOR_HOURLY =
  'TOGGLE_ACTIVITY_CONTRACTOR_HOURLY';
export const TOGGLE_ACTIVITY_SECTION = 'TOGGLE_ACTIVITY_SECTION';
export const UPDATE_ACTIVITY = 'UPDATE_ACTIVITY';

const actionWithYears = (type, other) => (dispatch, getState) =>
  dispatch({ type, ...other, years: getState().apd.data.years });

export const addActivity = () => (dispatch, getState) => {
  const oldActivities = new Set(getState().activities.allKeys);

  dispatch(actionWithYears(ADD_ACTIVITY));
  dispatch(updateBudget());

  const newKey = getState().activities.allKeys.find(k => !oldActivities.has(k));

  dispatch({
    type: ADD_ACTIVITY_DIRTY,
    data: getState().activities.byKey[newKey]
  });
};

export const updateActivity = (key, updates, isExpense = false) => dispatch => {
  dispatch({
    type: UPDATE_ACTIVITY,
    key,
    updates
  });
  if (isExpense) {
    dispatch(updateBudget());
  }
};

export const addActivityContractor = (key, { save = saveApd } = {}) => async (
  dispatch,
  getState
) => {
  const activity = getState().activities.byKey[key];
  const oldIds = activity.contractorResources.map(c => c.id);

  dispatch(actionWithYears(ADD_ACTIVITY_CONTRACTOR, { key }));

  const newApd = await dispatch(save());
  const newContractor = newApd.activities
    .find(a => a.id === activity.id)
    .contractorResources.find(c => !oldIds.includes(c.id));

  const updates = {
    contractorResources: {
      [activity.contractorResources.length]: { id: newContractor.id }
    }
  };

  dispatch(updateActivity(key, updates));
};

export const addActivityGoal = key => ({ type: ADD_ACTIVITY_GOAL, key });

export const addActivityExpense = key =>
  actionWithYears(ADD_ACTIVITY_EXPENSE, { key });

export const addActivityMilestone = key => ({
  type: ADD_ACTIVITY_MILESTONE,
  key
});

export const addActivityStatePerson = key =>
  actionWithYears(ADD_ACTIVITY_STATE_PERSON, { key });

export const expandActivitySection = key => ({
  type: EXPAND_ACTIVITY_SECTION,
  key
});

export const removeActivity = (key, { global = window } = {}) => dispatch => {
  if (global.confirm('Do you really want to delete this activity?')) {
    dispatch({ type: REMOVE_ACTIVITY, key });
    dispatch(updateBudget());
  }
};

export const removeActivityContractor = (
  key,
  contractorKey,
  { global = window } = {}
) => dispatch => {
  if (
    global.confirm('Do you really want to delete this contractor resource?')
  ) {
    dispatch({
      type: REMOVE_ACTIVITY_CONTRACTOR,
      key,
      contractorKey
    });
    dispatch(updateBudget());
  }
};

export const removeActivityGoal = (
  key,
  goalKey,
  { global = window } = {}
) => dispatch => {
  if (global.confirm('Do you really want to delete this activity goal?')) {
    dispatch({
      type: REMOVE_ACTIVITY_GOAL,
      key,
      goalKey
    });
  }
};

export const removeActivityExpense = (
  key,
  expenseKey,
  { global = window } = {}
) => dispatch => {
  if (global.confirm('Do you really want to delete this non-personnel cost?')) {
    dispatch({
      type: REMOVE_ACTIVITY_EXPENSE,
      key,
      expenseKey
    });
    dispatch(updateBudget());
  }
};

export const removeActivityMilestone = (
  key,
  milestoneKey,
  { global = window } = {}
) => dispatch => {
  if (global.confirm('Do you really want to delete this activity milestone?')) {
    dispatch({
      type: REMOVE_ACTIVITY_MILESTONE,
      key,
      milestoneKey
    });
  }
};

export const removeActivityStatePerson = (
  key,
  personKey,
  { global = window } = {}
) => dispatch => {
  if (global.confirm('Do you really want to delete this personnel entry?')) {
    dispatch({
      type: REMOVE_ACTIVITY_STATE_PERSON,
      key,
      personKey
    });
    dispatch(updateBudget());
  }
};

export const toggleActivitySection = key => ({
  type: TOGGLE_ACTIVITY_SECTION,
  key
});

export const toggleActivityContractorHourly = (
  key,
  contractorKey,
  useHourly
) => dispatch => {
  dispatch({
    type: TOGGLE_ACTIVITY_CONTRACTOR_HOURLY,
    key,
    contractorKey,
    useHourly
  });
  dispatch(updateBudget());
};
