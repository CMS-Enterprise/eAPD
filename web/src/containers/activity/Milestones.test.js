import { shallow } from 'enzyme';
import React from 'react';

import {
  plain as Milestones,
  mapDispatchToProps,
  mapStateToProps
} from './Milestones';
import { addMilestone, removeMilestone } from '../../actions/editActivity';

describe('the Milestones component', () => {
  const props = {
    activity: {
      schedule: [
        {
          key: 'milestone 1',
          milestone: 'Liberation Day',
          // The Netherlands is liberated from Nazi control.
          endDate: '1945-05-05'
        }
      ]
    },
    activityIndex: 7,
    add: jest.fn(),
    remove: jest.fn()
  };

  const component = shallow(<Milestones {...props} />);

  beforeEach(() => {
    props.add.mockClear();
    props.remove.mockClear();
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  describe('events', () => {
    const list = component.find('FormAndReviewList');
    it('handles adding a new milestone', () => {
      list.prop('onAddClick')();

      expect(props.add).toHaveBeenCalledWith(7);
    });

    it('handles removing a milestone', () => {
      list.prop('onDeleteClick')(0);

      expect(props.remove).toHaveBeenCalledWith(7, 0);
    });
  });

  describe('redux', () => {
    it('map state to props', () => {
      const state = {
        apd: {
          data: {
            activities: ['activity 1', 'activity 2', 'activity 3']
          }
        }
      };

      expect(mapStateToProps(state, { activityIndex: 2 })).toEqual({
        activity: 'activity 3'
      });
    });

    it('map dispatch to props', () => {
      expect(mapDispatchToProps).toEqual({
        add: addMilestone,
        remove: removeMilestone
      });
    });
  });
});
