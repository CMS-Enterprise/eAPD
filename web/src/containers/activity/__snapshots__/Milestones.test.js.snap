// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Milestones component renders correctly 1`] = `
<Subsection
  headerClassName={null}
  id={null}
  nested={false}
  resource="activities.milestones"
>
  <div
    className="mb3"
  >
    <hr />
    <FormAndReviewList
      activityIndex={7}
      addButtonText="Add another milestone"
      allowDeleteAll={false}
      className={null}
      collapsed={[Function]}
      expanded={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(MilestoneForm)",
          "type": [Function],
        }
      }
      extraItemButtons={Array []}
      list={
        Array [
          Object {
            "endDate": "1945-05-05",
            "key": "milestone 1",
            "milestone": "Liberation Day",
          },
        ]
      }
      noDataMessage="Add milestones for this activity."
      onAddClick={[Function]}
      onDeleteClick={[Function]}
    />
  </div>
</Subsection>
`;
